backend/seafile/seafile.go

```diff
@@ -0,0 +1,1144 @@
+	encrypted           bool         // Is this an encrypted library
+	rootDirectory       string       // directory part of root (if any)
+	opt                 Options      // parsed options
+	libraries           *cache.Cache // Keep a cache of libraries
+	librariesMutex      sync.Mutex   // Mutex to protect getLibraryID
+	createLibraryMutex  sync.Mutex   // Mutex to protect library creation
+	features            *fs.Features // optional features
+	endpoint            *url.URL     // URL of the host
+	endpointURL         string       // endpoint as a string
+	srv                 *rest.Client // the connection to the one drive server
+	pacer               *fs.Pacer    // pacer for API calls
+	authMu              sync.Mutex   // Mutex to protect library decryption
+	rangeDownloadNotice sync.Once    // Display the notice only once
+	useOldDirectoryAPI  bool         // Use the old API v2 if seafile < 7
+}
+
+// ------------------------------------------------------------
+
+// NewFs constructs an Fs from the path, container:path
+func NewFs(name, root string, m configmap.Mapper) (fs.Fs, error) {
+	// Parse config into Options struct
+	opt := new(Options)
+	err := configstruct.Set(m, opt)
+	if err != nil {
+		return nil, err
+	}
+	root = strings.Trim(root, "/")
+	isLibraryRooted := opt.LibraryName != ""
+	var libraryName, rootDirectory string
+	if isLibraryRooted {
+		libraryName = opt.LibraryName
+		rootDirectory = root
+	} else {
+		libraryName, rootDirectory = bucket.Split(root)
+	}
+
+	if !strings.HasSuffix(opt.URL, "/") {
+		opt.URL += "/"
+	}
+	if opt.Password != "" {
+		var err error
+		opt.Password, err = obscure.Reveal(opt.Password)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't decrypt user password")
+		}
+	}
+	if opt.LibraryKey != "" {
+		var err error
+		opt.LibraryKey, err = obscure.Reveal(opt.LibraryKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't decrypt library password")
+		}
+	}
+
+	// Parse the endpoint
+	u, err := url.Parse(opt.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Fs{
+		name:          name,
+		root:          root,
+		libraryName:   libraryName,
+		rootDirectory: rootDirectory,
+		libraries:     cache.New(),
+		opt:           *opt,
+		endpoint:      u,
+		endpointURL:   u.String(),
+		srv:           rest.NewClient(fshttp.NewClient(fs.Config)).SetRoot(u.String()),
+		pacer:         fs.NewPacer(pacer.NewDefault(pacer.MinSleep(minSleep), pacer.MaxSleep(maxSleep), pacer.DecayConstant(decayConstant))),
+	}
+	f.features = (&fs.Features{
+		CanHaveEmptyDirectories: true,
+		BucketBased:             opt.LibraryName == "",
+	}).Fill(f)
+
+	ctx := context.Background()
+	serverInfo, err := f.getServerInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	fs.Debugf(nil, "Seafile server version %s", serverInfo.Version)
+
+	// We don't support bellow seafile v6.0 (version 6.0 is already more than 3 years old)
+	serverVersion := semver.New(serverInfo.Version)
```

Comment:
Just a quick note: If this is the only place this is used, is there really a need to add yet another dependency to rclone? I'm a big fan of reducing dependency bloat. I realize that this is a very small and simple library but for something that could be done with one simple regex this still feels kinda unnecessary for me.

Reply:
That is a reasonable point, especially if we only use the major version.

Conversely I note we already have quite a lot of semvar parsing code already here:

https://github.com/rclone/rclone/blob/master/fs/version/version.go

Which might be a good candidate for replacement by that library...

Reply:
Yes this is a good point, I never noticed the version class in the code.
I could update to use the in-house code, or alternatively replace the in-house implementation by the semver library?